include/async/algorithm.hpp

```diff
@@ -412,16 +410,34 @@ struct race_and_cancel_operation<Receiver, frg::tuple<Functors...>, std::index_s
 
 public:
 	race_and_cancel_operation(race_and_cancel_sender<Functors...> s, Receiver r)
-	: r_{std::move(r)}, ops_{make_operations_tuple(std::move(s))}, n_done_{0} { }
+	: r_{std::move(r)}, ops_{make_operations_tuple(std::move(s))}, n_sync_{0}, n_done_{0} { }
 
-	void start() {
-		(execution::start_inline(ops_.template get<Is>()), ...);
+	bool start_inline() {
+		unsigned int n_sync = 0;
+
+		((execution::start_inline(ops_.template get<Is>())
+			? n_sync++ : 0), ...);
+
+		auto n = n_done_.fetch_add(n_sync, std::memory_order_acq_rel);
+
+		if (!n && n_sync) {
+			for(unsigned int j = 0; j < sizeof...(Is); ++j)
+				cs_[j].cancel();
+		}
```

Comment:
This can be put into a `if(n_sync)` block to avoid an atomic op if no operations finish synchronously.